File: PART_iOS/ViewHierarchy/ViewHierarchyObjC/ViewHierarchyObjC/ViewController.m
//
//  Copyright (c) 2016 Keun young Kim <book@meetkei.com>
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.
//

#import "ViewController.h"

@interface ViewController ()
@property (strong, nonatomic) UILabel *redView;
@property (strong, nonatomic) UILabel* greenView;
@property (strong, nonatomic) UILabel *blueView;
@end

@implementation ViewController

- (UILabel*)makeGreenView {
    CGRect frame = self.redView.frame;
    frame.origin.x = 50;
    frame.origin.y = 50;
    
    UILabel* gView = [[UILabel alloc] initWithFrame:frame];
    gView.backgroundColor = [UIColor greenColor];
    gView.textColor = [UIColor blackColor];
    gView.font = [UIFont boldSystemFontOfSize:80];
    gView.textAlignment = NSTextAlignmentCenter;
    gView.text = @"G";
    
    return gView;
}

- (IBAction)addGreenView:(id)sender {
    if (self.greenView == nil) {
        UILabel* gView = [self makeGreenView];
        self.greenView = gView;
    }
    
    [self.view addSubview:self.greenView];
}

- (IBAction)removeGreenView:(id)sender {
    if (self.greenView != nil) {
        [self.greenView removeFromSuperview];
        self.greenView = nil;
    }
}

- (IBAction)addGreenViewAboveRedView:(id)sender {
    if (self.greenView == nil) {
        UILabel* gView = [self makeGreenView];
        self.greenView = gView;
    }
    
    [self.view insertSubview:self.greenView aboveSubview:self.redView];
}

- (IBAction)addGreenViewBelowRedView:(id)sender {
    if (self.greenView == nil) {
        UILabel* gView = [self makeGreenView];
        self.greenView = gView;
    }
    
    [self.view insertSubview:self.greenView belowSubview:self.redView];
}

- (IBAction)bringGreenViewToFront:(id)sender {
    if ([self.greenView isDescendantOfView:self.view]) {
        [self.view bringSubviewToFront:self.greenView];
    }
}

- (IBAction)sendGreenViewToBack:(id)sender {
    if ([self.greenView isDescendantOfView:self.view]) {
        [self.view sendSubviewToBack:self.greenView];
    }
}

- (IBAction)exchangeRedViewWithGreenView:(id)sender {
    NSUInteger gIndex = NSNotFound;
    NSUInteger rIndex = NSNotFound;
    
    NSArray* list = self.view.subviews;
    for (UIView* sv in list) {
        if (sv == self.greenView) {
            gIndex = [list indexOfObject:sv];
        } else if (sv == self.redView) {
            rIndex = [list indexOfObject:sv];
        }
    }
    
    if (gIndex != NSNotFound && rIndex != NSNotFound) {
        [self.view exchangeSubviewAtIndex:gIndex withSubviewAtIndex:rIndex];
    }
}

- (IBAction)addGreenViewToBlueView:(id)sender {
    if (self.greenView != nil) {
        [self.blueView addSubview:self.greenView];
    }
}

- (IBAction)toggleClipsToBounds:(id)sender {
    self.blueView.clipsToBounds = !self.blueView.clipsToBounds;
}


- (IBAction)addGreenViewToRootView:(id)sender {
    [self addGreenView:sender];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Point #1
    CGRect frame = CGRectMake(20, 20, 100, 100);
    UILabel* rView = [[UILabel alloc] initWithFrame:frame];
    rView.backgroundColor = [UIColor redColor];
    rView.textColor = [UIColor whiteColor];
    rView.font = [UIFont boldSystemFontOfSize:80];
    rView.textAlignment = NSTextAlignmentCenter;
    rView.text = @"R";
    
    // Point #2
    [self.view addSubview:rView];
    self.redView = rView;
    
    // Point #3
    frame = CGRectMake(80, 80, 100, 100);
    UILabel* bView = [[UILabel alloc] initWithFrame:frame];
    bView.backgroundColor = [UIColor blueColor];
    bView.textColor = [UIColor whiteColor];
    bView.font = [UIFont boldSystemFontOfSize:80];
    bView.textAlignment = NSTextAlignmentCenter;
    bView.text = @"B";
    
    [self.view addSubview:bView];
    self.blueView = bView;
}

@end

File: PART_iOS/ViewHierarchy/ViewHierarchySwift2/ViewHierarchySwift2/ViewController.swift
//
//  Copyright (c) 2016 Keun young Kim <book@meetkei.com>
//
//  Permission is hereby granted, free of charge, to any person obtaining a copy
//  of this software and associated documentation files (the "Software"), to deal
//  in the Software without restriction, including without limitation the rights
//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
//  copies of the Software, and to permit persons to whom the Software is
//  furnished to do so, subject to the following conditions:
//
//  The above copyright notice and this permission notice shall be included in
//  all copies or substantial portions of the Software.
//
//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
//  THE SOFTWARE.
//

import UIKit

class ViewController: UIViewController {
    
    lazy var redView: UILabel = {
        let frame = CGRect(x: 20, y: 20, width: 100, height: 100)
        let label = UILabel(frame: frame)
        label.backgroundColor = UIColor.redColor()
        label.textColor = UIColor.whiteColor()
        label.font = UIFont.boldSystemFontOfSize(80)
        label.textAlignment = .Center
        label.text = "R"
        
        return label
    }()
    
    lazy var blueView: UILabel = {
        let frame = CGRect(x: 80, y: 80, width: 100, height: 100)
        let label = UILabel(frame: frame)
        label.backgroundColor = UIColor.blueColor()
        label.textColor = UIColor.whiteColor()
        label.font = UIFont.boldSystemFontOfSize(80)
        label.textAlignment = .Center
        label.text = "B"
        
        return label
    }()
    
    var greenView: UILabel?
    
    func makeGreenView() -> UILabel {
        var frame = redView.frame
        frame.origin.x = 50
        frame.origin.y = 50
        
        let gView = UILabel(frame: frame)
        gView.backgroundColor = UIColor.greenColor()
        gView.textColor = UIColor.blackColor()
        gView.font = UIFont.boldSystemFontOfSize(80)
        gView.textAlignment = .Center
        gView.text = "G"
        
        return gView
    }
    
    @IBAction func addGreenView(sender: AnyObject) {
        if greenView == nil {
            greenView = makeGreenView()
        }
        
        if let gView = greenView {
            view.addSubview(gView)
        }
    }

    @IBAction func removeGreenView(sender: AnyObject) {
        if let gView = greenView {
            gView.removeFromSuperview()
            greenView = nil
        }
    }
    
    @IBAction func addGreenViewAboveRedView(sender: AnyObject) {
        if greenView == nil {
            greenView = makeGreenView()
        }
        
        if let gView = greenView {
            view.insertSubview(gView, aboveSubview: redView)
        }
    }
    
    @IBAction func addGreenViewBelowRedView(sender: AnyObject) {
        if greenView == nil {
            greenView = makeGreenView()
        }
        
        if let gView = greenView {
            view.insertSubview(gView, belowSubview: redView)
        }
    }
    
    @IBAction func bringGreenViewToFront(sender: AnyObject) {
        if let gView = greenView where gView.isDescendantOfView(view) {
            view.bringSubviewToFront(gView)
        }
    }
    
    @IBAction func sendGreenViewToBack(sender: AnyObject) {
        if let gView = greenView where gView.isDescendantOfView(view) {
            view.sendSubviewToBack(gView)
        }
    }
    
    @IBAction func exchangeRedViewWithGreenView(sender: AnyObject) {
        guard let gView = greenView, let gIndex = view.subviews.indexOf(gView) else {
            return
        }
        
        guard let rIndex = view.subviews.indexOf(redView) else {
            return
        }
        
        view.exchangeSubviewAtIndex(gIndex, withSubviewAtIndex: rIndex)
    }
    
    @IBAction func addGreenViewToBlueView(sender: AnyObject) {
        if let gView = greenView {
            blueView.addSubview(gView)
        }
    }
    
    @IBAction func toggleClipsToBounds(sender: AnyObject) {
        blueView.clipsToBounds = !blueView.clipsToBounds
    }
    
    @IBAction func addGreenViewToRootView(sender: AnyObject) {
        addGreenView(sender)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        view.addSubview(redView)
        view.addSubview(blueView)
    }
}





